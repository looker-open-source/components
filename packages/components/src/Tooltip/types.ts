/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { Transitions } from '@looker/design-tokens'
import { Placement } from '@popperjs/core'
import { Property } from 'csstype'
import { MouseEvent, ReactElement, ReactNode } from 'react'
import { MenuDomProps } from '../Menu'

// import { UsePopoverResponseDom } from '../Popover'

export type TooltipRenderProp = (props: UseTooltipResponseDom) => ReactNode

export interface UseTooltipProps {
  /**
   * Specify a callback to be called before trying to close the Tooltip. This allows for
   * use-cases where the user might lose work (think common "Save before closing warning" type flow)
   * Specify a callback to be called each time this Tooltip is closed
   */
  canClose?: () => boolean
  isOpen?: boolean
  /**
   * Can be one of: top, bottom, left, right, auto, with the modifiers: start,
   * end. This value comes directly from popperjs. See
   * https://popper.js.org/popper-documentation.html#Popper.placements for more
   * info.
   */
  placement?: Placement
  /**
   * Content to display inside the tooltip. Can be a string or JSX.
   * If not defined, the Tooltip will not render.
   */
  content?: ReactNode
  /**
   * Specify a fixed content width.
   * @default auto
   */
  width?: string
  /**
   * Specify a fixed max-width.
   * @default none
   */
  maxWidth?: string
  /**
   * Specify the text alignment within tooltips.
   * @default center
   */
  textAlign?: Property.TextAlign

  /**
   * The id of the span containing the tooltip text (if absent, a random id will be generated)
   */
  id?: string

  /**
   * Tooltip background and text color is inverted from the page default
   * @default true
   * @private
   */
  invert?: boolean

  /**
   * The trigger element ref to use (if absent, one will be created and returned)
   */
  triggerElement?: HTMLElement | null

  /**
   * If true, the useTooltip hook will return nothing
   */
  disabled?: boolean

  /**
   * Delay
   */
  delay?: Transitions
}

export interface UseTooltipResponseDom {
  'aria-describedby': string
  className: string
  onBlur: () => void
  onFocus: () => void
  onMouseOut: (event: MouseEvent<HTMLElement>) => void
  onMouseOver: () => void
  /**
   * Used by popper.js to position the OverlaySurface relative to the trigger
   */
  ref: (node: HTMLElement | null) => void
}
export interface TooltipProps extends UseTooltipProps, Partial<MenuDomProps> {
  content: ReactNode
  /**
   * Component to receive tooltip behavior or render prop function that
   * receives tooltip props and returns a component
   */
  children:
    | ReactElement<UseTooltipResponseDom & MenuDomProps>
    | TooltipRenderProp
}
