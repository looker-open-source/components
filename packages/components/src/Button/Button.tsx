/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import { ButtonBase } from './ButtonBase'

export const Button = styled(ButtonBase)`
  background: ${({ theme, color = 'key' }) => theme.colors[color]};
  border: 1px solid ${({ theme, color = 'key' }) => theme.colors[color]};
  color: ${({ theme, color = 'key' }) => theme.colors[`${color}Text`]};

  &:hover,
  &:focus,
  &.hover {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
  }

  &[aria-expanded='true'],
  &:active,
  &.active {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
  }

  &[disabled] {
    &:hover,
    &:active,
    &:focus {
      background-color: ${({ theme, color = 'key' }) => theme.colors[color]};
      border-color: ${({ theme }) => theme.colors.ui3};
    }
  }
`
