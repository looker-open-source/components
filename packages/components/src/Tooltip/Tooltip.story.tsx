/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { FormEvent, SyntheticEvent, useState } from 'react'
import { Story } from '@storybook/react/types-6-0'
import { Button } from '../Button'
import { Card } from '../Card'
import { FieldToggleSwitch } from '../Form'
import { Space, SpaceVertical } from '../Layout'
import { Text } from '../Text'
import { Popover, PopoverContent } from '../Popover'
import { Tooltip, TooltipProps } from './'

const Template: Story<TooltipProps> = (args) => (
  <Tooltip {...args}>
    <Button m="xxxlarge">Open Tooltip</Button>
  </Tooltip>
)

export const Basic = Template.bind({})
Basic.args = {
  content: 'Simple Content',
}
Basic.parameters = {
  storyshots: { disable: true },
}

export const Open = Template.bind({})
Open.args = {
  ...Basic.args,
  isOpen: true,
}
Open.parameters = {
  storyshots: { disable: true },
}

export const PlacemenTop = Template.bind({})
PlacemenTop.args = {
  ...Basic.args,
  placement: 'top',
}
PlacemenTop.parameters = {
  storyshots: { disable: true },
}

export const PlacementRight = Template.bind({})
PlacementRight.args = {
  ...Basic.args,
  placement: 'right',
}
PlacementRight.parameters = {
  storyshots: { disable: true },
}

export const PlacementLeft = Template.bind({})
PlacementLeft.args = {
  ...Basic.args,
  content: 'Left',
  placement: 'left',
}
PlacementLeft.parameters = {
  storyshots: { disable: true },
}

export const DelayNone = Template.bind({})
DelayNone.args = {
  ...Basic.args,
  delay: 'none',
}
DelayNone.parameters = {
  storyshots: { disable: true },
}

export const OpenDelayNone = Template.bind({})
OpenDelayNone.args = {
  ...DelayNone.args,
  isOpen: true,
}
OpenDelayNone.parameters = {
  storyshots: { disable: true },
}

export const RenderProp = () => (
  <Tooltip content="Start editing" placement="right">
    {(tooltipProps) => (
      <Button m="xxxlarge" {...tooltipProps}>
        Open Tooltip
      </Button>
    )}
  </Tooltip>
)

RenderProp.parameters = {
  storyshots: { disable: true },
}

export const LargeTrigger = () => (
  <Tooltip content="See what happens when you scroll" placement="right">
    <Card width={500} height={800} raised p="large">
      Very large trigger
    </Card>
  </Tooltip>
)

LargeTrigger.parameters = {
  storyshots: { disable: true },
}

export const NestedInPopover = () => {
  const [prevent, setPrevent] = useState(false)
  function handleChange(e: FormEvent<HTMLInputElement>) {
    setPrevent(e.currentTarget.checked)
  }

  const [lastEvent, setLastEvent] = useState('N/A')
  const getHandler = (text: string) => (e: SyntheticEvent) => {
    setLastEvent(text)
    if (prevent) {
      e.preventDefault()
    }
  }

  const handlers = {
    onBlur: getHandler('blur'),
    onClick: getHandler('click'),
    onFocus: getHandler('focus'),
    onMouseOut: getHandler('mouse out'),
    onMouseOver: getHandler('mouse over'),
  }

  return (
    <SpaceVertical p="large">
      <Text>Last event: {lastEvent}</Text>
      <Space>
        <Popover content={<PopoverContent>Some content</PopoverContent>}>
          <Tooltip content="Some tooltip">
            <Button {...handlers}>Open</Button>
          </Tooltip>
        </Popover>
        <FieldToggleSwitch
          on={prevent}
          onChange={handleChange}
          label="Prevent default"
        />
      </Space>
    </SpaceVertical>
  )
}

NestedInPopover.parameters = {
  storyshots: { disable: true },
}

export default {
  component: Tooltip,
  title: 'Tooltip',
}
