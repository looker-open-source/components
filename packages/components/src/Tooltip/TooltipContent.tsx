import styled from 'styled-components'
import { Link } from '../Link'
import { Paragraph } from '../Text'
import { TooltipProps } from './Tooltip'
export const TooltipContent = styled(Paragraph).attrs(
  (props: TooltipProps) => ({
    fontSize: 'xsmall',
    lineHeight: 'xsmall',
    m: 'none',
    maxWidth: props.width,
    p: 'xsmall',
    width: 'auto',
  })
)`
  white-space: normal;
  text-transform: none;
  word-break: break-word;
  hyphens: auto;
  overflow-wrap: anywhere;

  ${Link} {
    color: ${(props) => props.theme.colors.palette.blue200};
    text-decoration: underline;

    &:focus,
    &:hover {
      color: ${(props) => props.theme.colors.palette.blue100};
    }
  }
`

TooltipContent.defaultProps = { textAlign: 'center', width: '16rem' }
