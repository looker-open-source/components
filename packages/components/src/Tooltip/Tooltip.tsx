/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { CustomizableAttributes } from '@looker/design-tokens'
import { TextAlignProperty } from 'csstype'
import { Placement } from '@popperjs/core'
import React, {
  cloneElement,
  FC,
  isValidElement,
  useMemo,
  useState,
  ReactNode,
  Ref,
} from 'react'
import omit from 'lodash/omit'
import { ModalContext } from '../Modal'
import {
  useCallbackRef,
  useID,
  usePopper,
  UsePopperProps,
  useForkedRef,
} from '../utils'
import { OverlaySurface, SurfaceStyleProps } from '../Overlay/OverlaySurface'
import { TooltipContent } from './TooltipContent'

export interface UseTooltipProps {
  /**
   * Display and arrow that points to the trigger element on popovers
   * @default true
   */
  arrow?: boolean

  /**
   * Specify a callback to be called before trying to close the Modal. This allows for
   * use-cases where the user might lose work (think common "Save before closing warning" type flow)
   * Specify a callback to be called each time this Modal is closed
   */
  canClose?: () => boolean

  isOpen?: boolean
  /**
   * Can be one of: top, bottom, left, right, auto, with the modifiers: start,
   * end. This value comes directly from popperjs. See
   * https://popper.js.org/popper-documentation.html#Popper.placements for more
   * info.
   */
  placement?: Placement
  /**
   * Content to display inside the tooltip. Can be a string or JSX.
   */
  content?: ReactNode
  /**
   * Specify a fixed content width.
   * @default auto
   */
  width?: string
  /**
   * Specify the text alignment within tooltips.
   * @default center
   */
  textAlign?: TextAlignProperty

  /**
   * The id of the span containing the tooltip text (if absent, a random id will be generated)
   */
  id?: string

  /**
   * The trigger element ref to use (if absent, one will be created and returned)
   */
  triggerElement?: HTMLElement | null

  /**
   * If true, the useTooltip hook will return nothing
   */
  disabled?: boolean

  /**
   * Customizes the style of the tooltip
   */
  surfaceStyles?: SurfaceStyleProps
}

export const CustomizableTooltipAttributes: CustomizableAttributes = {}

type TooltipRenderProp = (tooltipProps: {
  'aria-describedby': string
  onBlur: () => void
  onFocus: () => void
  onMouseOut: (event: React.MouseEvent<Element, MouseEvent>) => void
  onMouseOver: () => void
  ref: Ref<any>
}) => ReactNode

export interface TooltipProps extends UseTooltipProps {
  content: ReactNode
  /**
   * Component to wrap. The HOC will listen for mouse events on this component, maintain the
   * state of isOpen accordingly, and pass that state into the children or "trigger" element
   */
  children: ReactNode | TooltipRenderProp
}
export function useTooltip({
  arrow = true,
  canClose,
  content,
  isOpen: initializeOpen = false,
  width,
  textAlign,
  disabled,
  surfaceStyles,
  id,
  triggerElement,
  placement: propsPlacement = 'bottom',
}: UseTooltipProps) {
  const [isOpen, setIsOpen] = useState(initializeOpen)
  const [surfaceElement, surfaceCallbackRef] = useCallbackRef()
  const [newTriggerElement, callbackRef] = useCallbackRef()
  // If the triggerElement is passed in props, use that instead of the new element
  const element =
    typeof triggerElement === 'undefined' ? newTriggerElement : triggerElement

  const handleOpen = () => setIsOpen(true)
  const handleClose = () => {
    if (canClose && !canClose()) return
    setIsOpen(false)
  }

  const handleMouseOut = (event: React.MouseEvent) => {
    if (!isOpen) return

    const related = event.relatedTarget

    if (element && (element === related || element.contains(related as Node))) {
      return
    }

    if (
      surfaceElement &&
      (surfaceElement === related || surfaceElement.contains(related as Node))
    ) {
      return
    }

    window.requestAnimationFrame(() => {
      handleClose()
    })
  }

  const usePopperProps: UsePopperProps = useMemo(
    () => ({
      anchor: element,
      arrow,
      options: {
        modifiers: [
          {
            enabled: true,
            name: 'flip',
            options: {
              flipVariations: true,
            },
          },
        ],
        placement: propsPlacement,
      },
    }),
    [arrow, element, propsPlacement]
  )
  const {
    arrowProps,
    placement,
    popperInstanceRef,
    style,
    targetRef,
  } = usePopper(usePopperProps)

  const ref = useForkedRef(targetRef, surfaceCallbackRef)

  const guaranteedId = useID(id)

  const popper =
    isOpen && content && !disabled ? (
      <ModalContext.Provider value={{ closeModal: handleClose }}>
        <OverlaySurface
          arrow={arrow}
          arrowProps={arrowProps}
          eventHandlers={{ onMouseOut: handleMouseOut }}
          placement={placement}
          ref={ref}
          style={style}
          zIndex={CustomizableTooltipAttributes.zIndex}
          backgroundColor="inverse"
          borderRadius="medium"
          boxShadow={3}
          color="inverseOn"
          {...surfaceStyles}
        >
          <TooltipContent
            role="tooltip"
            id={guaranteedId}
            width={width}
            textAlign={textAlign}
          >
            {content}
          </TooltipContent>
        </OverlaySurface>
      </ModalContext.Provider>
    ) : null

  return {
    'aria-describedby': guaranteedId,
    isOpen,
    onBlur: handleClose,
    onFocus: handleOpen,
    onMouseOut: handleMouseOut,
    onMouseOver: handleOpen,
    popperInstanceRef,
    ref: callbackRef,
    tooltip: popper,
  }
}

function isRenderProp(
  children: ReactNode | TooltipRenderProp
): children is TooltipRenderProp {
  return typeof children === 'function'
}

export const Tooltip: FC<TooltipProps> = ({ children, ...props }) => {
  const tooltipProps = useTooltip(props)

  const tooltipPropsLabeled = {
    ...omit(tooltipProps, ['tooltip', 'popperInstanceRef', 'isOpen']),
    className: tooltipProps.isOpen ? 'hover' : '',
  }

  let target = children

  if (isValidElement(children)) {
    target = cloneElement(children, { ...tooltipPropsLabeled })
  } else if (isRenderProp(children)) {
    target = children(tooltipPropsLabeled)
  } else {
    // eslint-disable-next-line no-console
    console.warn(
      `Element "${typeof target}" can't be used as target for Tooltip`
    )
  }

  return (
    <>
      {tooltipProps.tooltip}
      {target}
    </>
  )
}
