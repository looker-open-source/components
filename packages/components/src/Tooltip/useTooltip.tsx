/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { Property } from 'csstype'
import { Placement } from '@popperjs/core'
import { Transitions } from '@looker/design-tokens'
import React, { MouseEvent, ReactNode, useMemo, useState } from 'react'
import {
  useAnimationState,
  useCallbackRef,
  useID,
  usePopper,
  UsePopperProps,
  useForkedRef,
} from '../utils'
import { Portal } from '../Portal'
import { TooltipContent } from './TooltipContent'
import { TooltipSurface } from './TooltipSurface'

export interface UseTooltipProps {
  /**
   * Specify a callback to be called before trying to close the Tooltip. This allows for
   * use-cases where the user might lose work (think common "Save before closing warning" type flow)
   * Specify a callback to be called each time this Tooltip is closed
   */
  canClose?: () => boolean
  isOpen?: boolean
  /**
   * Can be one of: top, bottom, left, right, auto, with the modifiers: start,
   * end. This value comes directly from popperjs. See
   * https://popper.js.org/popper-documentation.html#Popper.placements for more
   * info.
   */
  placement?: Placement
  /**
   * Content to display inside the tooltip. Can be a string or JSX.
   */
  content?: ReactNode
  /**
   * Specify a fixed content width.
   * @default auto
   */
  width?: string
  /**
   * Specify a fixed max-width.
   * @default none
   */
  maxWidth?: string
  /**
   * Specify the text alignment within tooltips.
   * @default center
   */
  textAlign?: Property.TextAlign

  /**
   * The id of the span containing the tooltip text (if absent, a random id will be generated)
   */
  id?: string

  /**
   * Tooltip background and text color is inverted from the page default
   * @default true
   * @private
   */
  invert?: boolean

  /**
   * The trigger element ref to use (if absent, one will be created and returned)
   */
  triggerElement?: HTMLElement | null

  /**
   * If true, the useTooltip hook will return nothing
   */
  disabled?: boolean

  /**
   * Delay
   */
  delay?: Transitions
}

export interface UseTooltipResponseDom {
  'aria-describedby': string
  className: string
  onBlur: () => void
  onFocus: () => void
  onMouseOut: (event: MouseEvent<HTMLElement>) => void
  onMouseOver: () => void
  /**
   * Used by popper.js to position the OverlaySurface relative to the trigger
   */
  ref: (node: HTMLElement | null) => void
}

export function useTooltip({
  canClose,
  content,
  isOpen: initializeOpen = false,
  width,
  maxWidth = '30rem',
  textAlign,
  disabled,
  id,
  invert,
  triggerElement,
  placement: propsPlacement = 'bottom',
  delay = 'intricate',
}: UseTooltipProps) {
  const [isOpen, setIsOpen] = useState(initializeOpen)
  const { busy, className, renderDOM } = useAnimationState(
    isOpen,
    delay,
    'none'
  )

  const [surfaceElement, surfaceCallbackRef] = useCallbackRef()
  const [newTriggerElement, callbackRef] = useCallbackRef()
  // If the triggerElement is passed in props, use that instead of the new element
  const element =
    typeof triggerElement === 'undefined' ? newTriggerElement : triggerElement

  const handleOpen = () => {
    if (!element || !element.dataset.notooltip) {
      setIsOpen(true)
    }
  }
  const handleClose = () => {
    if (canClose && !canClose()) return
    setIsOpen(false)
  }

  const handleMouseOut = (event: MouseEvent<HTMLElement>) => {
    if (!isOpen) return

    const related = event.relatedTarget

    if (element && (element === related || element.contains(related as Node))) {
      return
    }

    if (
      surfaceElement &&
      (surfaceElement === related || surfaceElement.contains(related as Node))
    ) {
      return
    }

    window.requestAnimationFrame(() => {
      handleClose()
    })
  }

  const usePopperProps: UsePopperProps = useMemo(
    () => ({
      anchor: element,
      options: {
        modifiers: [
          {
            enabled: true,
            name: 'flip',
            options: {
              flipVariations: true,
            },
          },
        ],
        placement: propsPlacement,
      },
    }),
    [element, propsPlacement]
  )
  const { placement, popperInstanceRef, style, targetRef } = usePopper(
    usePopperProps
  )

  const ref = useForkedRef(targetRef, surfaceCallbackRef)

  const guaranteedId = useID(id)

  const popper =
    renderDOM && content && !disabled ? (
      <Portal>
        <TooltipSurface
          aria-busy={busy ? true : undefined}
          className={className}
          eventHandlers={{ onMouseOut: handleMouseOut }}
          placement={placement}
          ref={ref}
          style={style}
          maxWidth={maxWidth}
          invert={invert}
        >
          <TooltipContent
            role="tooltip"
            id={guaranteedId}
            width={width}
            textAlign={textAlign}
          >
            {content}
          </TooltipContent>
        </TooltipSurface>
      </Portal>
    ) : null

  return {
    domProps: {
      'aria-describedby': guaranteedId,
      className: renderDOM ? 'hover' : '',
      onBlur: handleClose,
      onFocus: handleOpen,
      onMouseOut: handleMouseOut,
      onMouseOver: handleOpen,
      ref: callbackRef,
    },
    popperInstanceRef,
    tooltip: popper,
  }
}
