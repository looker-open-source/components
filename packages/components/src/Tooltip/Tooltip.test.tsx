/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import { renderWithTheme } from '@looker/components-test-utils'
import { fireEvent, screen } from '@testing-library/react'
import { Button } from '../Button'
import { Tooltip } from './Tooltip'

describe('Tooltip', () => {
  let rafSpy: jest.SpyInstance<number, [FrameRequestCallback]>

  beforeEach(() => {
    rafSpy = jest
      .spyOn(window, 'requestAnimationFrame')
      .mockImplementation((cb: any) => cb())
  })

  afterEach(() => {
    rafSpy.mockRestore()
  })

  test('trigger: open on mouseover, close on mouseout', () => {
    renderWithTheme(
      <Tooltip content="Hello world" id="stable-id">
        <Button>Test</Button>
      </Tooltip>
    )

    const trigger = screen.getByText('Test')

    fireEvent.mouseOver(trigger)

    const tooltip = screen.getByText('Hello world')
    expect(tooltip.parentNode).toMatchSnapshot()

    fireEvent.mouseOut(trigger)
    expect(tooltip).not.toBeInTheDocument()
  })

  test('close on surface mouseout', () => {
    renderWithTheme(
      <Tooltip content="Hello world" isOpen>
        <Button>Test</Button>
      </Tooltip>
    )

    const trigger = screen.getByText('Test')

    fireEvent.mouseOver(trigger)

    const tooltip = screen.getByText('Hello world')
    expect(tooltip).toBeVisible()

    fireEvent.mouseOut(tooltip)
    expect(tooltip).not.toBeInTheDocument()
  })

  test('open initially, collapse on mouseout', () => {
    renderWithTheme(
      <Tooltip content="Hello world" isOpen>
        <Button>Test</Button>
      </Tooltip>
    )

    const trigger = screen.getByText('Test')
    const tooltip = screen.queryByText('Hello world')
    expect(tooltip).toBeVisible()

    fireEvent.mouseOut(trigger)
    expect(tooltip).not.toBeInTheDocument()
  })

  test('supports styling props', () => {
    renderWithTheme(
      <Tooltip content="Hello world" width="20rem" textAlign="right">
        <Button>Test</Button>
      </Tooltip>
    )

    const trigger = screen.getByText('Test')

    fireEvent.mouseOver(trigger)

    const tooltip = screen.queryByText('Hello world')
    expect(tooltip).toBeVisible()

    expect(tooltip).toHaveStyle('max-width: 20rem')
    expect(tooltip).toHaveStyle('text-align: right')
    fireEvent.mouseOut(trigger)
  })

  test('Render props version works', () => {
    renderWithTheme(
      <Tooltip content="Hello world">
        {(props) => <Button {...props}>Test</Button>}
      </Tooltip>
    )

    const trigger = screen.getByText('Test')

    fireEvent.mouseOver(trigger)

    const tooltip = screen.queryByText('Hello world')
    expect(tooltip).toBeVisible()

    fireEvent.mouseOut(trigger)
    expect(tooltip).not.toBeInTheDocument()
  })
})
