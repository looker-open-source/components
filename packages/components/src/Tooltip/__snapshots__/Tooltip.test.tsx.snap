// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip trigger: open on mouseover, close on mouseout 1`] = `
.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  border-radius: 0.25rem;
  box-shadow: 0 3px 18px rgba(0,0,0,0.12),0 1px 4px rgba(0,0,0,0.04);
  background-color: #262D33;
  color: #FFFFFF;
}

.c0:focus {
  outline: none;
}

.c1 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  font-size: 0.75rem;
  line-height: 1rem;
  text-align: center;
  margin: 0rem;
  padding: 0.5rem;
  width: auto;
  max-width: 16rem;
  -webkit-hyphens: auto;
  -moz-hyphens: auto;
  -ms-hyphens: auto;
  hyphens: auto;
  overflow-wrap: anywhere;
  text-transform: none;
  white-space: normal;
  word-break: break-word;
}

.c1 .c2 {
  color: #E8E5FF;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 .c2:focus,
.c1 .c2:hover {
  color: #F3F2FF;
}

.c1 .c2:active {
  color: #FFFFFF;
}

<div
  class="c0"
  color="inverseOn"
>
  <p
    class="c1"
    font-size="xsmall"
    id="stable-id"
    role="tooltip"
    width="auto"
  >
    Hello world
  </p>
</div>
`;
